src/bun.js/node/net/BlockList.zig

```diff
@@ -0,0 +1,264 @@
+
+const js = JSC.Codegen.JSBlockList;
+pub const fromJS = js.fromJS;
+pub const toJS = js.toJS;
+
+ref_count: RefCount = .init(),
+globalThis: *JSC.JSGlobalObject,
+da_rules: std.ArrayList(Rule),
+mutex: std.Thread.Mutex = .{},
+
+pub fn constructor(globalThis: *JSC.JSGlobalObject, callFrame: *JSC.CallFrame) bun.JSError!*@This() {
+    _ = callFrame;
+    const ptr = @This().new(.{
+        .globalThis = globalThis,
+        .da_rules = .init(bun.default_allocator),
+    });
+    return ptr;
+}
+
+pub fn estimatedSize(this: *@This()) usize {
+    this.mutex.lock();
+    defer this.mutex.unlock();
+    return @sizeOf(@This()) + (@sizeOf(Rule) * this.da_rules.items.len);
+}
+
+pub fn finalize(this: *@This()) void {
+    this.deref();
+}
+
+pub fn deinit(this: *@This()) void {
+    this.da_rules.deinit();
+    bun.destroy(this);
+}
+
+pub fn isBlockList(globalThis: *JSC.JSGlobalObject, callframe: *JSC.CallFrame) bun.JSError!JSC.JSValue {
+    _ = globalThis;
+    const value = callframe.argumentsAsArray(1)[0];
+    return .jsBoolean(value.as(@This()) != null);
+}
+
+// [kInspect](depth, options)
+pub fn customInspect(this: *@This(), globalThis: *JSC.JSGlobalObject, callframe: *JSC.CallFrame) bun.JSError!JSC.JSValue {
+    _ = this;
+    _ = globalThis;
+    _ = callframe;
+    @compileError("TODO");
+}
+
+pub fn addAddress(this: *@This(), globalThis: *JSC.JSGlobalObject, callframe: *JSC.CallFrame) bun.JSError!JSC.JSValue {
+    this.mutex.lock();
+    defer this.mutex.unlock();
+    const arguments = callframe.argumentsAsArray(2);
+    const address_js, var family_js = arguments;
+    var new_address = false;
+    if (family_js.isUndefined()) family_js = bun.String.static("ipv4").toJS(globalThis);
+    const address = address_js.as(SocketAddress) orelse blk: {
+        new_address = true;
+        try validators.validateString(globalThis, address_js, "address", .{});
+        try validators.validateString(globalThis, family_js, "family", .{});
+        break :blk try SocketAddress.createFromAddrFamily(globalThis, address_js, family_js);
+    };
+    defer if (new_address) bun.destroy(address);
+    try this.da_rules.insert(0, .{ .addr = address._addr });
+    return .jsUndefined();
+}
+
+pub fn addRange(this: *@This(), globalThis: *JSC.JSGlobalObject, callframe: *JSC.CallFrame) bun.JSError!JSC.JSValue {
+    this.mutex.lock();
+    defer this.mutex.unlock();
+    const arguments = callframe.argumentsAsArray(3);
+    const start_js, const end_js, var family_js = arguments;
+    var new_start = false;
+    var new_end = false;
+    if (family_js.isUndefined()) family_js = bun.String.static("ipv4").toJS(globalThis);
+    const start = start_js.as(SocketAddress) orelse blk: {
+        new_start = true;
+        try validators.validateString(globalThis, start_js, "start", .{});
+        try validators.validateString(globalThis, family_js, "family", .{});
+        break :blk try SocketAddress.createFromAddrFamily(globalThis, start_js, family_js);
```

Comment:
This works fine but it looks like it should be possible to make a parse function that returns `_addr` directly that could avoid allocating a SocketAddress at all, and `fn create()` would be updated to be `return .new(.{ ._addr = parse(), ._presentation = options.address })`

Then, this line would then be `const start = if(start_js.as(SocketAddress)) |sock_addr| sock_addr._addr else {  ... break :blk parse()  }`

Reply:
oh good point